File: Syncfusion_Sample/Windows/Grid.Windows/Samples/2.0/Product Showcase/Excel Like UI Demo/CS/FormulaRangeSelectionController.cs
#region Copyright Syncfusion Inc. 2001 - 2007
//
//  Copyright Syncfusion Inc. 2001 - 2007. All rights reserved.
//
//  Use of this code is subject to the terms of our license.
//  A copy of the current license can be obtained at any time by e-mailing
//  licensing@syncfusion.com. Any infringement will be prosecuted under
//  applicable laws. 
//
#endregion

using System;
using System.ComponentModel;
using System.Diagnostics;
using System.Drawing;
using System.Drawing.Drawing2D;
using System.Windows.Forms;

using Syncfusion.Diagnostics;
using Syncfusion.Windows.Forms;
using Syncfusion.Windows.Forms.Grid;

namespace ExcelLikeUI_2005
{

	/// <summary>
	/// Summary description for FormulaRangeSelectionController.
	/// </summary>
	public class FormulaRangeSelectionController : IMouseController
	{
		//
		GridControlBase grid;
		GridRangeInfo activeRange;
		int lastHitTestCode = GridHitTestContext.None;
		const int selectingCellsHitTestCode = 101;

		//used for public properties
		private bool mouseRequiresControlKey = false;
		private bool keyRequiresControlKey = false;
		private Color selectColor = Color.LightBlue;
		private bool resetSelectAfterPlacement = true;

		//used to highlight the formula range selection as it changes
		private DragWindow dragWindow;
		
		public FormulaRangeSelectionController(GridControlBase grid)
		{
			this.grid = grid;

			//events used to react to keyboard actions
			this.grid.CurrentCellKeyDown += new KeyEventHandler(grid_CurrentCellKeyDown);
			this.grid.CurrentCellDeactivated += new GridCurrentCellDeactivatedEventHandler(grid_CurrentCellDeactivated);
			this.grid.CurrentCellValidateString += new GridCurrentCellValidateStringEventHandler(grid_CurrentCellValidateString);
		}

		#region Properties specific to FormulaRangeSelectionController

		/// <summary>
		/// Indicates whether the user needs to press control while
		/// selecting the range with the mouse.
		/// </summary>
		public bool MouseRequiresControlKey
		{
			get{return mouseRequiresControlKey;}
			set{mouseRequiresControlKey = value;}
		}

		/// <summary>
		/// Indicates whether the user needs to press control while
		/// selecting the range with the arrow keys.
		/// </summary>
		public bool KeyBoardRequiresControlKey
		{
			get{return keyRequiresControlKey;}
			set{keyRequiresControlKey = value;}
		}

		/// <summary>
		/// Color used to mark the cells being selected.
		/// </summary>
		public Color SelectionColor
		{
			get{return selectColor;}
			set{selectColor = value;}
		}

		/// <summary>
		/// Indicates whether the placed text should remain selected when control is returned
		/// to the user after selecting cells.
		/// </summary>
		public bool ResetSelectionAfterPlacement
		{
			get{return resetSelectAfterPlacement;}
			set{resetSelectAfterPlacement = value;}
		}
		#endregion

		#region base implementation properties
		public string Name 
		{ 
			get	{return "FormulaRangeSelector";}
		}

		public Cursor Cursor 
		{ 
			get	{return Cursors.Default;}
		}
		
		#endregion

		#region keyboard handling for selecting cells with arrow keys

		//indicates whether cells are being selected with the keyboard
		private bool inKeyMove = false;

		//holds the location of the most recently key selected cell
		private int keyRow = -1;
		private int keyCol = -1;

		//handle keys not handled by the active textbox
		//arrow keys, escape, enter, ...
		private void grid_CurrentCellKeyDown(object sender, KeyEventArgs e)
		{
			GridCurrentCell cc = this.grid.CurrentCell;
					
			if(e.KeyCode == Keys.Escape)
			{	//cancel selection mode
				if(inKeyMove)
				{
					e.Handled = true;
					TextBoxBase gtb = (TextBoxBase) cc.Renderer.Control;
					gtb.SelectedText = ""; //clear the text
				}
				CancelMode();
			}
			else
			{   //only check keys for active editing formula cell with cursor in OK position
				if(cc.HasControlFocus && cc.Renderer is GridFormulaCellRenderer
					&& CanDoKeySelection(cc))
				{
					bool handled = false;
					switch (e.KeyCode)
					{
						case Keys.Up:
							if(inKeyMove)
								keyRow -= (keyRow > this.grid.Model.Rows.HeaderCount) ? 1 : 0;
							else
							{
								keyRow = cc.RowIndex - ((cc.RowIndex > this.grid.Model.Rows.HeaderCount) ? 1 : 0);
								keyCol = cc.ColIndex;
							}
							handled = true;
							break;
						case Keys.Down:
							if(inKeyMove)
								keyRow += (keyRow < this.grid.Model.RowCount) ? 1 : 0;
							else
							{
								keyRow = cc.RowIndex + ((cc.RowIndex < this.grid.Model.RowCount) ? 1 : 0);
								keyCol = cc.ColIndex;
							}
							handled = true;
							break;
						case Keys.Left:
							if(inKeyMove)
								keyCol -= (keyCol > this.grid.Model.Rows.HeaderCount) ? 1 : 0;
							else
							{
								keyCol = cc.ColIndex - ((cc.ColIndex > this.grid.Model.Rows.HeaderCount) ? 1 : 0);
								keyRow = cc.RowIndex;
							}
							handled = true;
							break;
						case Keys.Right:
							if(inKeyMove)
								keyCol += (keyCol < this.grid.Model.ColCount) ? 1 : 0;
							else
							{
								keyCol = cc.ColIndex + ((cc.ColIndex < this.grid.Model.ColCount) ? 1 : 0);
								keyRow = cc.RowIndex;
							}
							handled = true;
							break;
						case Keys.Enter:
							if(inKeyMove)
							{
								inKeyMove = false;
								e.Handled = true;
								PlaceTextInCell(ResetSelectionAfterPlacement);
								CancelMode();
								return;
							}
						
							break;
						default:
							break;
					}
					if(handled)
					{
						if(inKeyMove && 0 != (Control.ModifierKeys & Keys.Shift))
						{
							this.MoveDragToRowCol(keyRow, keyCol);
						}
						else
						{
							CancelMode();
							this.StartDragAtRowCol(keyRow, keyCol);
							inKeyMove = true;
							//puts text in cell
							this.MoveDragToRowCol(keyRow, keyCol);
						}
					}
					else
						inKeyMove = false;
					e.Handled = handled;
				}
			}
			
		}

		//turn off everything when the current cell deactivates
		private void grid_CurrentCellDeactivated(object sender, GridCurrentCellDeactivatedEventArgs e)
		{
			CancelMode();
		}

		//catches the keypress of a key normally handled by the textbox
		//here we start the range slection and append the keystroke
		private void grid_CurrentCellValidateString(object sender, GridCurrentCellValidateStringEventArgs e)
		{
			string s = e.Text;
			int len = s.Length;
			if(inKeyMove && len > 0)
			{
				GridCurrentCell cc = this.grid.CurrentCell;
				TextBoxBase gtb = (TextBoxBase) cc.Renderer.Control;
				gtb.Text = gtb.Text + s[len - 1];
				gtb.SelectionStart = gtb.Text.Length + 1;
				gtb.SelectionLength = 0;
				CancelMode();
				e.Cancel = true;
			}
		}
		#endregion

		#region Text Management during a selection

		private int initDragRow = -1;
		private int intDragCol = -1;
		private void StartDragAtRowCol(int row, int col)
		{
			activeRange = GridRangeInfo.Cell(row, col);

			initDragRow = row;
			intDragCol = col;
					
			dragWindow = new DragWindow();
			dragWindow.Opacity = .4f;
			dragWindow.BackColor = SelectionColor;
			dragWindow.Bounds = GetBounds(activeRange);
			dragWindow.ShowWindowTopMost();
		}
		
		private void PlaceTextInCell(bool resetSelection)
		{
			string range = GridRangeInfo.GetAlphaLabel(activeRange.Left) + GridRangeInfo.GetNumericLabel(activeRange.Top);
			if(activeRange.Top != activeRange.Bottom || activeRange.Left != activeRange.Right)
			{
				range += ':' + GridRangeInfo.GetAlphaLabel(activeRange.Right) + GridRangeInfo.GetNumericLabel(activeRange.Bottom);
			}

			TextBoxBase gtb = GetTextBox();
			
			int start = gtb.SelectionStart;

			gtb.SelectedText = range;
			if(resetSelection)
			{
				gtb.SelectionStart = start + range.ToString().Length;
				gtb.SelectionLength = 0;
			}
			else
			{	
				gtb.SelectionStart = start;
				gtb.SelectionLength = range.ToString().Length;
			}
		}

		private void MoveDragToRowCol(int row, int col)
		{
			if(row > grid.Model.Rows.FrozenCount)
				row = Math.Max(grid.TopRowIndex, row);
			if(col > grid.Model.Cols.FrozenCount)
				col = Math.Max(grid.LeftColIndex, col);

			
			int top = Math.Min(initDragRow, row);
			int left = Math.Min(intDragCol, col);
			int bottom = Math.Max(initDragRow, row);
			int right = Math.Max(intDragCol, col);
			activeRange = GridRangeInfo.Cells(top, 
				left,
				bottom, 
				right);
			PlaceTextInCell(false);
			dragWindow.Bounds = GetBounds(activeRange);
			dragWindow.ShowWindowTopMost();
		}

		#endregion

		#region Code to allow the start of selections

		private string validPrecedingChars = "+-/*<>=(,:";
		public string ValidPrecedingChars
		{
			get{return validPrecedingChars;}
			set{validPrecedingChars = value;}
		}

		protected virtual bool CanDoSelection(GridCurrentCell cc)
		{ 
			TextBoxBase gtb = GetTextBox();
			string s = gtb.Text;
			bool b = s.Length > 0  && gtb.SelectionStart > 0 
				&& validPrecedingChars.IndexOf(s[gtb.SelectionStart-1]) > -1;
			return b;
		}

		protected virtual bool CanDoKeySelection(GridCurrentCell cc)
		{ 
			bool ctrl = (Control.ModifierKeys & Keys.Control) != 0;

			bool b = CanDoSelection(cc) || this.inKeyMove;
			
			if(b)
			{
				TextBoxBase gtb = (TextBoxBase) cc.Renderer.Control;
				string s = gtb.Text;
				b = this.inKeyMove 
					|| (KeyBoardRequiresControlKey && ctrl)
					|| (!KeyBoardRequiresControlKey && gtb.SelectionStart == gtb.Text.Length);
			}
		 
			return b;
		}
		#endregion

		#region base implementation overrides
		public void MouseHoverEnter()
		{
		}

		/// <summary>
		/// User is moving the mouse over the hot-test area
		/// </summary>
		/// <param name="e"></param>
		public void MouseHover(MouseEventArgs e)
		{
		}

		/// <summary>
		/// Called when the hovering ends, either when user has moved mouse away from hittest area
		/// or when the user has pressed a mouse button.
		/// </summary>
		public void MouseHoverLeave(EventArgs e)
		{
		}

		Control savedFocus = null;

		public void MouseDown(MouseEventArgs e)
		{
			
			if (e.Button == MouseButtons.Left)
			{
				if (e.Clicks == 1)
				{
					// Start automatic scrolling when user drags mouse out of window
					grid.AutoScrolling = ScrollBars.Both;
					
					// 
					int rowIndex, colIndex;
					Point point = new Point(e.X, e.Y);
					grid.PointToRowCol(point, out rowIndex, out colIndex);
					StartDragAtRowCol(rowIndex, colIndex);

					savedFocus = grid.FindForm().ActiveControl;
				}
			}
			
			// DoubleClick
		}

		



		Rectangle GetBounds(GridRangeInfo range)
		{
			Rectangle bounds = grid.RangeInfoToRectangle(range, GridRangeOptions.None);
			bounds.Intersect(grid.ClientRectangle);
			bounds = grid.RectangleToScreen(bounds);
			return bounds;
		}

		/// <summary>
		/// User has dragged mouse. If mouse is down, set current position.
		/// </summary>
		/// <param name="e"></param>
		public void MouseMove(MouseEventArgs e)
		{
			// Set current position
			int rowIndex, colIndex;
			
			Point point = new Point(e.X, e.Y);
			try
			{
				grid.PointToRowCol(point, out rowIndex, out colIndex, 0);
				MoveDragToRowCol(rowIndex, colIndex);
			}
			catch
			{
			}
		}

		/// <summary>
		/// User has release mouse button. Stop automatic scrolling.
		/// </summary>
		/// <param name="e"></param>
		public void MouseUp(MouseEventArgs e)
		{
			// Stop automatic scrolling 
			grid.AutoScrolling = ScrollBars.None;

			int rowIndex, colIndex;
			Point point = new Point(e.X, e.Y);
			grid.PointToRowCol(point, out rowIndex, out colIndex);

			dragWindow.Visible = false;
			dragWindow = null;
			
			PlaceTextInCell(ResetSelectionAfterPlacement);
			grid.FocusOnMouseDown = true;
			savedFocus = null;
		}



		public void CancelMode()
		{
			grid.AutoScrolling = ScrollBars.None;
			if(dragWindow != null)
				dragWindow.Visible = false;
			dragWindow = null;
			inKeyMove = false;
			grid.FocusOnMouseDown = true;

			savedFocus = null;
		}

		GridCurrentCell GetCurrentCell()
		{
			GridCurrentCell cc = null;
			if (grid.Parent is SplitterControl)
			{
				Control c = ((SplitterControl) grid.Parent).ActivePane;
				if (c is GridControlBase)
					cc = ((GridControlBase) c).CurrentCell;
			}
			if (cc == null)
			{
				if (grid.Model.ActiveGridView != null)
					cc = grid.Model.ActiveGridView.CurrentCell;
				if (cc == null)
					cc = grid.CurrentCell;
			}
			return cc;
		}

		TextBoxBase GetTextBox()
		{
			if (savedFocus is TextBoxBase)
				return (TextBoxBase) savedFocus;
			else if (GetCurrentCell().Renderer.Control is TextBoxBase)
				return (TextBoxBase) GetCurrentCell().Renderer.Control;
			return null;
		}

		
		public int HitTest(MouseEventArgs e, IMouseController controller)
		{
			grid.FocusOnMouseDown = true; // avoid Focus() being called 
			lastHitTestCode = GridHitTestContext.None;
			GridCurrentCell cc = GetCurrentCell();
			if(dragWindow == null)
			{
				if(cc.IsEditing && cc.Renderer is GridFormulaCellRenderer)
				{
					bool ctrl = (Control.ModifierKeys & Keys.Control) != 0;
				
					if (e.Button == MouseButtons.Left && (!MouseRequiresControlKey || ctrl))
					{
						Point point = new Point(e.X, e.Y);
						int rowIndex, colIndex;
						grid.PointToRowCol(point, out rowIndex, out colIndex);

						if(rowIndex != cc.RowIndex || colIndex != cc.ColIndex)
						{
							if(CanDoSelection(cc))
							{
								lastHitTestCode = selectingCellsHitTestCode;
								grid.FocusOnMouseDown = false;
							}
						}
					
					}
				}
			}
			return lastHitTestCode;
		}
		#endregion


	}
}

File: Syncfusion_Sample/Windows/Grid.Windows/Samples/2.0/Product Showcase/Excel Like UI Demo/VB/FormulaRangeSelectionController.vb
#Region "Copyright Syncfusion Inc. 2001 - 2007"
'
'  Copyright Syncfusion Inc. 2001 - 2007. All rights reserved.
'
'  Use of this code is subject to the terms of our license.
'  A copy of the current license can be obtained at any time by e-mailing
'  licensing@syncfusion.com. Any infringement will be prosecuted under
'  applicable laws. 
'
#End Region


Imports Microsoft.VisualBasic
Imports System
Imports System.ComponentModel
Imports System.Diagnostics
Imports System.Drawing
Imports System.Drawing.Drawing2D
Imports System.Windows.Forms

Imports Syncfusion.Diagnostics
Imports Syncfusion.Windows.Forms
Imports Syncfusion.Windows.Forms.Grid

Namespace ExcelLikeUI_2005

	''' <summary>
	''' Summary description for FormulaRangeSelectionController.
	''' </summary>
	Public Class FormulaRangeSelectionController
		Implements IMouseController
		'
		Private grid As GridControlBase
		Private activeRange As GridRangeInfo
		Private lastHitTestCode As Integer = GridHitTestContext.None
		Private Const selectingCellsHitTestCode As Integer = 101

		'used for public properties
		Private mouseRequiresControlKey_Renamed As Boolean = False
		Private keyRequiresControlKey As Boolean = False
		Private selectColor As Color = Color.LightBlue
		Private resetSelectAfterPlacement As Boolean = True

		'used to highlight the formula range selection as it changes
		Private dragWindow As DragWindow

		Public Sub New(ByVal grid As GridControlBase)
			Me.grid = grid

			'events used to react to keyboard actions
			AddHandler grid.CurrentCellKeyDown, AddressOf grid_CurrentCellKeyDown
			AddHandler grid.CurrentCellDeactivated, AddressOf grid_CurrentCellDeactivated
			AddHandler grid.CurrentCellValidateString, AddressOf grid_CurrentCellValidateString
		End Sub

		#Region "Properties specific to FormulaRangeSelectionController"

		''' <summary>
		''' Indicates whether the user needs to press control while
		''' selecting the range with the mouse.
		''' </summary>
		Public Property MouseRequiresControlKey() As Boolean
			Get
				Return mouseRequiresControlKey_Renamed
			End Get
			Set(ByVal value As Boolean)
				mouseRequiresControlKey_Renamed = value
			End Set
		End Property

		''' <summary>
		''' Indicates whether the user needs to press control while
		''' selecting the range with the arrow keys.
		''' </summary>
		Public Property KeyBoardRequiresControlKey() As Boolean
			Get
				Return keyRequiresControlKey
			End Get
			Set(ByVal value As Boolean)
				keyRequiresControlKey = value
			End Set
		End Property

		''' <summary>
		''' Color used to mark the cells being selected.
		''' </summary>
		Public Property SelectionColor() As Color
			Get
				Return selectColor
			End Get
			Set(ByVal value As Color)
				selectColor = value
			End Set
		End Property

		''' <summary>
		''' Indicates whether the placed text should remain selected when control is returned
		''' to the user after selecting cells.
		''' </summary>
		Public Property ResetSelectionAfterPlacement() As Boolean
			Get
				Return resetSelectAfterPlacement
			End Get
			Set(ByVal value As Boolean)
				resetSelectAfterPlacement = value
			End Set
		End Property
		#End Region

		#Region "base implementation properties"
		Public ReadOnly Property Name() As String Implements IMouseController.Name
			Get
				Return "FormulaRangeSelector"
			End Get
		End Property

		Public ReadOnly Property Cursor() As Cursor Implements IMouseController.Cursor
			Get
				Return Cursors.Default
			End Get
		End Property

		#End Region

		#Region "keyboard handling for selecting cells with arrow keys"

		'indicates whether cells are being selected with the keyboard
		Private inKeyMove As Boolean = False

		'holds the location of the most recently key selected cell
		Private keyRow As Integer = -1
		Private keyCol As Integer = -1

		'handle keys not handled by the active textbox
		'arrow keys, escape, enter, ...
		Private Sub grid_CurrentCellKeyDown(ByVal sender As Object, ByVal e As KeyEventArgs)
			Dim cc As GridCurrentCell = Me.grid.CurrentCell

			If e.KeyCode = Keys.Escape Then
				If inKeyMove Then
					e.Handled = True
					Dim gtb As TextBoxBase = CType(cc.Renderer.Control, TextBoxBase)
					gtb.SelectedText = "" 'clear the text
				End If
				CancelMode()
			Else
				If cc.HasControlFocus AndAlso TypeOf cc.Renderer Is GridFormulaCellRenderer AndAlso CanDoKeySelection(cc) Then
					Dim handled As Boolean = False
					Select Case e.KeyCode
						Case Keys.Up
							If inKeyMove Then
								If (keyRow > Me.grid.Model.Rows.HeaderCount) Then
									keyRow -= 1
								Else
									keyRow -= 0
								End If
							Else
								If (cc.RowIndex > Me.grid.Model.Rows.HeaderCount) Then
									keyRow = cc.RowIndex - (1)
								Else
									keyRow = cc.RowIndex - (0)
								End If
								keyCol = cc.ColIndex
							End If
							handled = True
						Case Keys.Down
							If inKeyMove Then
								If (keyRow < Me.grid.Model.RowCount) Then
									keyRow += 1
								Else
									keyRow += 0
								End If
							Else
								If (cc.RowIndex < Me.grid.Model.RowCount) Then
									keyRow = cc.RowIndex + (1)
								Else
									keyRow = cc.RowIndex + (0)
								End If
								keyCol = cc.ColIndex
							End If
							handled = True
						Case Keys.Left
							If inKeyMove Then
								If (keyCol > Me.grid.Model.Rows.HeaderCount) Then
									keyCol -= 1
								Else
									keyCol -= 0
								End If
							Else
								If (cc.ColIndex > Me.grid.Model.Rows.HeaderCount) Then
									keyCol = cc.ColIndex - (1)
								Else
									keyCol = cc.ColIndex - (0)
								End If
								keyRow = cc.RowIndex
							End If
							handled = True
						Case Keys.Right
							If inKeyMove Then
								If (keyCol < Me.grid.Model.ColCount) Then
									keyCol += 1
								Else
									keyCol += 0
								End If
							Else
								If (cc.ColIndex < Me.grid.Model.ColCount) Then
									keyCol = cc.ColIndex + (1)
								Else
									keyCol = cc.ColIndex + (0)
								End If
								keyRow = cc.RowIndex
							End If
							handled = True
						Case Keys.Enter
							If inKeyMove Then
								inKeyMove = False
								e.Handled = True
								PlaceTextInCell(ResetSelectionAfterPlacement)
								CancelMode()
								Return
							End If

						Case Else
					End Select
					If handled Then
						If inKeyMove AndAlso 0 <> (Control.ModifierKeys And Keys.Shift) Then
							Me.MoveDragToRowCol(keyRow, keyCol)
						Else
							CancelMode()
							Me.StartDragAtRowCol(keyRow, keyCol)
							inKeyMove = True
							'puts text in cell
							Me.MoveDragToRowCol(keyRow, keyCol)
						End If
					Else
						inKeyMove = False
					End If
					e.Handled = handled
				End If
			End If

		End Sub

		'turn off everything when the current cell deactivates
		Private Sub grid_CurrentCellDeactivated(ByVal sender As Object, ByVal e As GridCurrentCellDeactivatedEventArgs)
			CancelMode()
		End Sub

		'catches the keypress of a key normally handled by the textbox
		'here we start the range slection and append the keystroke
		Private Sub grid_CurrentCellValidateString(ByVal sender As Object, ByVal e As GridCurrentCellValidateStringEventArgs)
			Dim s As String = e.Text
			Dim len As Integer = s.Length
			If inKeyMove AndAlso len > 0 Then
				Dim cc As GridCurrentCell = Me.grid.CurrentCell
				Dim gtb As TextBoxBase = CType(cc.Renderer.Control, TextBoxBase)
				gtb.Text = gtb.Text & s.Chars(len - 1)
				gtb.SelectionStart = gtb.Text.Length + 1
				gtb.SelectionLength = 0
				CancelMode()
				e.Cancel = True
			End If
		End Sub
		#End Region

		#Region "Text Management during a selection"

		Private initDragRow As Integer = -1
		Private intDragCol As Integer = -1
		Private Sub StartDragAtRowCol(ByVal row As Integer, ByVal col As Integer)
			activeRange = GridRangeInfo.Cell(row, col)

			initDragRow = row
			intDragCol = col

			dragWindow = New DragWindow()
			dragWindow.Opacity =.4f
			dragWindow.BackColor = SelectionColor
			dragWindow.Bounds = GetBounds(activeRange)
			dragWindow.ShowWindowTopMost()
		End Sub

		Private Sub PlaceTextInCell(ByVal resetSelection As Boolean)
			Dim range As String = GridRangeInfo.GetAlphaLabel(activeRange.Left) + GridRangeInfo.GetNumericLabel(activeRange.Top)
			If activeRange.Top <> activeRange.Bottom OrElse activeRange.Left <> activeRange.Right Then
				range &= ":"c + GridRangeInfo.GetAlphaLabel(activeRange.Right) + GridRangeInfo.GetNumericLabel(activeRange.Bottom)
			End If

			Dim gtb As TextBoxBase = GetTextBox()

			Dim start As Integer = gtb.SelectionStart

			gtb.SelectedText = range
			If resetSelection Then
                gtb.SelectionStart = Convert.ToInt32(start & range.ToString().Length)
				gtb.SelectionLength = 0
			Else
				gtb.SelectionStart = start
				gtb.SelectionLength = range.ToString().Length
			End If
		End Sub

		Private Sub MoveDragToRowCol(ByVal row As Integer, ByVal col As Integer)
			If row > grid.Model.Rows.FrozenCount Then
				row = Math.Max(grid.TopRowIndex, row)
			End If
			If col > grid.Model.Cols.FrozenCount Then
				col = Math.Max(grid.LeftColIndex, col)
			End If


			Dim top As Integer = Math.Min(initDragRow, row)
			Dim left As Integer = Math.Min(intDragCol, col)
			Dim bottom As Integer = Math.Max(initDragRow, row)
			Dim right As Integer = Math.Max(intDragCol, col)
			activeRange = GridRangeInfo.Cells(top, left, bottom, right)
			PlaceTextInCell(False)
			dragWindow.Bounds = GetBounds(activeRange)
			dragWindow.ShowWindowTopMost()
		End Sub

		#End Region

		#Region "Code to allow the start of selections"

		Private validPrecedingChars_Renamed As String = "+-/*<>=(,:"
		Public Property ValidPrecedingChars() As String
			Get
				Return validPrecedingChars_Renamed
			End Get
			Set(ByVal value As String)
				validPrecedingChars_Renamed = value
			End Set
		End Property

		Protected Overridable Function CanDoSelection(ByVal cc As GridCurrentCell) As Boolean
			Dim gtb As TextBoxBase = GetTextBox()
			Dim s As String = gtb.Text
			Dim b As Boolean = s.Length > 0 AndAlso gtb.SelectionStart > 0 AndAlso validPrecedingChars_Renamed.IndexOf(s.Chars(gtb.SelectionStart-1)) > -1
			Return b
		End Function

		Protected Overridable Function CanDoKeySelection(ByVal cc As GridCurrentCell) As Boolean
			Dim ctrl As Boolean = (Control.ModifierKeys And Keys.Control) <> 0

			Dim b As Boolean = CanDoSelection(cc) OrElse Me.inKeyMove

			If b Then
				Dim gtb As TextBoxBase = CType(cc.Renderer.Control, TextBoxBase)
				Dim s As String = gtb.Text
				b = Me.inKeyMove OrElse (KeyBoardRequiresControlKey AndAlso ctrl) OrElse ((Not KeyBoardRequiresControlKey) AndAlso gtb.SelectionStart = gtb.Text.Length)
			End If

			Return b
		End Function
		#End Region

		#Region "base implementation overrides"
		Public Sub MouseHoverEnter() Implements IMouseController.MouseHoverEnter
		End Sub

		''' <summary>
		''' User is moving the mouse over the hot-test area
		''' </summary>
		''' <param name="e"></param>
		Public Sub MouseHover(ByVal e As MouseEventArgs) Implements IMouseController.MouseHover
		End Sub

		''' <summary>
		''' Called when the hovering ends, either when user has moved mouse away from hittest area
		''' or when the user has pressed a mouse button.
		''' </summary>
		Public Sub MouseHoverLeave(ByVal e As EventArgs) Implements IMouseController.MouseHoverLeave
		End Sub

		Private savedFocus As Control = Nothing

		Public Sub MouseDown(ByVal e As MouseEventArgs) Implements IMouseController.MouseDown

			If e.Button = MouseButtons.Left Then
				If e.Clicks = 1 Then
					' Start automatic scrolling when user drags mouse out of window
					grid.AutoScrolling = ScrollBars.Both

					' 
					Dim rowIndex, colIndex As Integer
					Dim point As Point = New Point(e.X, e.Y)
					grid.PointToRowCol(point, rowIndex, colIndex)
					StartDragAtRowCol(rowIndex, colIndex)

					savedFocus = grid.FindForm().ActiveControl
				End If
			End If

			' DoubleClick
		End Sub





		Private Function GetBounds(ByVal range As GridRangeInfo) As Rectangle
			Dim bounds As Rectangle = grid.RangeInfoToRectangle(range, GridRangeOptions.None)
			bounds.Intersect(grid.ClientRectangle)
			bounds = grid.RectangleToScreen(bounds)
			Return bounds
		End Function

		''' <summary>
		''' User has dragged mouse. If mouse is down, set current position.
		''' </summary>
		''' <param name="e"></param>
		Public Sub MouseMove(ByVal e As MouseEventArgs) Implements IMouseController.MouseMove
			' Set current position
			Dim rowIndex, colIndex As Integer

			Dim point As Point = New Point(e.X, e.Y)
			Try
				grid.PointToRowCol(point, rowIndex, colIndex, 0)
				MoveDragToRowCol(rowIndex, colIndex)
			Catch
			End Try
		End Sub

		''' <summary>
		''' User has release mouse button. Stop automatic scrolling.
		''' </summary>
		''' <param name="e"></param>
		Public Sub MouseUp(ByVal e As MouseEventArgs) Implements IMouseController.MouseUp
			' Stop automatic scrolling 
			grid.AutoScrolling = ScrollBars.None

			Dim rowIndex, colIndex As Integer
			Dim point As Point = New Point(e.X, e.Y)
			grid.PointToRowCol(point, rowIndex, colIndex)

			dragWindow.Visible = False
			dragWindow = Nothing

			PlaceTextInCell(ResetSelectionAfterPlacement)
			grid.FocusOnMouseDown = True
			savedFocus = Nothing
		End Sub



		Public Sub CancelMode() Implements IMouseController.CancelMode
			grid.AutoScrolling = ScrollBars.None
			If Not dragWindow Is Nothing Then
				dragWindow.Visible = False
			End If
			dragWindow = Nothing
			inKeyMove = False
			grid.FocusOnMouseDown = True

			savedFocus = Nothing
		End Sub

		Private Function GetCurrentCell() As GridCurrentCell
			Dim cc As GridCurrentCell = Nothing
			If TypeOf grid.Parent Is SplitterControl Then
				Dim c As Control = (CType(grid.Parent, SplitterControl)).ActivePane
				If TypeOf c Is GridControlBase Then
					cc = (CType(c, GridControlBase)).CurrentCell
				End If
			End If
			If cc Is Nothing Then
				If Not grid.Model.ActiveGridView Is Nothing Then
					cc = grid.Model.ActiveGridView.CurrentCell
				End If
				If cc Is Nothing Then
					cc = grid.CurrentCell
				End If
			End If
			Return cc
		End Function

		Private Function GetTextBox() As TextBoxBase
			If TypeOf savedFocus Is TextBoxBase Then
				Return CType(savedFocus, TextBoxBase)
			ElseIf TypeOf GetCurrentCell().Renderer.Control Is TextBoxBase Then
				Return CType(GetCurrentCell().Renderer.Control, TextBoxBase)
			End If
			Return Nothing
		End Function


		Public Function HitTest(ByVal e As MouseEventArgs, ByVal controller As IMouseController) As Integer Implements IMouseController.HitTest
			grid.FocusOnMouseDown = True ' avoid Focus() being called
			lastHitTestCode = GridHitTestContext.None
			Dim cc As GridCurrentCell = GetCurrentCell()
			If dragWindow Is Nothing Then
				If cc.IsEditing AndAlso TypeOf cc.Renderer Is GridFormulaCellRenderer Then
					Dim ctrl As Boolean = (Control.ModifierKeys And Keys.Control) <> 0

					If e.Button = MouseButtons.Left AndAlso ((Not MouseRequiresControlKey) OrElse ctrl) Then
						Dim point As Point = New Point(e.X, e.Y)
						Dim rowIndex, colIndex As Integer
						grid.PointToRowCol(point, rowIndex, colIndex)

						If rowIndex <> cc.RowIndex OrElse colIndex <> cc.ColIndex Then
							If CanDoSelection(cc) Then
								lastHitTestCode = selectingCellsHitTestCode
								grid.FocusOnMouseDown = False
							End If
						End If

					End If
				End If
			End If
			Return lastHitTestCode
		End Function
		#End Region


	End Class
End Namespace